pilot/pkg/bootstrap/namespacecontroller.go

```diff
@@ -51,83 +51,77 @@ type NamespaceController struct {
 	getData func() map[string]string
 	client  corev1.CoreV1Interface
 
-	queue queue.Instance
-
 	// Controller and store for namespace objects
 	namespaceController cache.Controller
-	// Controller and store for ConfigMap objects
-	configMapController cache.Controller
+	namespaceStore      cache.Store
 }
 
 // NewNamespaceController returns a pointer to a newly constructed NamespaceController instance.
 func NewNamespaceController(data func() map[string]string, kubeClient kubernetes.Interface) *NamespaceController {
 	c := &NamespaceController{
 		getData: data,
 		client:  kubeClient.CoreV1(),
-		queue:   queue.NewQueue(time.Second),
 	}
 
-	configmapInformer := informer.NewFilteredConfigMapInformer(kubeClient, metav1.NamespaceAll, 0,
-		cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
-		func(options *metav1.ListOptions) {
-			options.LabelSelector = fields.SelectorFromSet(configMapLabel).String()
-		})
-	configmapInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
-		UpdateFunc: func(oldObj, newObj interface{}) {
-			c.queue.Push(func() error {
-				return c.configMapChange(newObj)
-			})
-		},
-		DeleteFunc: func(obj interface{}) {
-			cm, ok := obj.(*v1.ConfigMap)
-			if !ok {
-				tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
-				if !ok {
-					log.Errorf("error decoding object, invalid type")
-					return
-				}
-				cm, ok = tombstone.Obj.(*v1.ConfigMap)
-				if !ok {
-					log.Errorf("error decoding object tombstone, invalid type")
-					return
-				}
-			}
-			c.queue.Push(func() error {
-				ns, err := kubeClient.CoreV1().Namespaces().Get(cm.Namespace, metav1.GetOptions{})
-				if err != nil {
-					return err
-				}
-				// If the namespace is terminating, we may get into a loop of trying to re-add the configmap back
-				// We should make sure the namespace still exists
-				if ns.Status.Phase != v1.NamespaceTerminating {
-					return c.insertDataForNamespace(cm.Namespace)
-				}
-				return nil
-			})
-		},
-	})
-	c.configMapController = configmapInformer
-
-	namespaceInformer := informer.NewNamespaceInformer(kubeClient, 0, cache.Indexers{})
-	namespaceInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
-		AddFunc: func(obj interface{}) {
-			c.queue.Push(func() error {
-				return c.namespaceChange(obj)
-			})
-		},
+	namespaceLW := listwatch.MultiNamespaceListerWatcher([]string{metav1.NamespaceAll}, func(namespace string) cache.ListerWatcher {
+		return &cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				return c.client.Namespaces().List(options)
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				return c.client.Namespaces().Watch(options)
+			}}
 	})
-	c.namespaceController = namespaceInformer
+
+	c.namespaceStore, c.namespaceController =
+		cache.NewInformer(namespaceLW, &v1.Namespace{}, namespaceResyncPeriod, cache.ResourceEventHandlerFuncs{
+			UpdateFunc: c.namespaceUpdated,
```

Comment:
why do we need to manage updates for a namespace?

Reply:
This update handler periodically writes the configmap to a namespace, which is useful when the controller needs to sync the configmaps in namespaces. If only adding events are handled, the controller may miss the adding events and cause the configmap to be out-of-sync. The following is an explanation of OnUpdate cited from go/pkg/mod/k8s.io/client-go@v0.18.0/tools/cache/controller.go.

//  OnUpdate is also called when a re-list happens, and it will
//      get called even if nothing changed. This is useful for periodically
//      evaluating or syncing something.